cylc/uiserver/resolvers.py

```diff
@@ -349,29 +349,37 @@ async def cat_log(cls, workflow: Tokens, info, task=None):
         # subprocess ends
         asyncio.create_task(cls.enqueue(proc.stdout, queue))
         op_id = info.root_value
+
         while True:
-            if op_id not in info.context['sub_statuses']:
-                break
-            if info.context['sub_statuses'][op_id] == 'stop':
+            # TODO detect if subscription has stopped?
+            if (
+                op_id in info.context['sub_statuses']
+                and info.context['sub_statuses'][op_id] == 'stop'
+            ):
+                print('$$$ stop')
                 with suppress(psutil.Error):
                     cat_log_proc = psutil.Process(proc.pid)
                     # cat-log process will terminate on tail termination
                     for tail_proc in cat_log_proc.children():
                         tail_proc.terminate()
                 # clean the subscription from the status dict
                 info.context['sub_statuses'].pop(op_id)
+                break
 
-            if queue.empty() and buffer:
-                yield list(buffer)
-                buffer.clear()
+            if queue.empty():
+                if buffer:
+                    print('yield')
+                    yield list(buffer)
+                    buffer.clear()
+                await asyncio.sleep(0.1)
             else:
                 line = await queue.get()
+                print('add-to-buffer', line)
                 buffer.append(line)
                 if len(buffer) >= 20:
                     yield list(buffer)
                     buffer.clear()
-
-            await asyncio.sleep(0.1)
+                    await asyncio.sleep(0)
```

Comment:
Changed the sleep logic so that it only sleeps once the buffer is full.

When we sleep here we only sleep for 0 seconds which just yields control so that other coroutines get to run.